Improve network names and grouping

jgosmann opened the pull request:
**Motivation and context:**
This is an attempt to improve the naming and grouping of modules; mainly for a better display in the Nengo GUI. 

### Naming
* All Nengo SPA modules get a default label. This label, unfortunately, takes precedence even when the module is assigned to a variable in Nengo GUI, i.e. `name = spa.State(64)` will display the name 'State' in Nengo GUI, not 'name'. The obvious workaround is `name = spa.State(64, label='name')`.
* Symbolic expression will display as such under the generated nodes (this was already the case prior to this PR). Example: `spa.sym.A * spa.sym.B * 0.3' will display as 'A*B*0.3'.
* Non-symbolic Semantic Pointer nodes will be label with 'Semantic Pointer (XXd)` where XX is the dimensionality.
* The utility nodes for basal ganglia inputs (the `ifmax` return value) are currently unlabeled. The user will not really interact with these as long as they are not assigned to a variable. So this make sense imo.

### Grouping

* Everything in the context of an action selection object gets now grouped into a network. I'm not sure whether that's better or worse. Unfortunately, this also includes all connections, so these won't be displayed in Nengo GUI until the network is 'opened'.
* Everything else gets shoved into the network where it is defined (as before). This allows the user to easily group things manually into separate networks. It, however, puts the connections again into the same network which can hide them in Nengo GUI.

Improvements on this grouping might hard because everything gets constructed immediately. For example if an `a*b` gets constructed, it is not known if it is part of `a*b >> state1` or `(a*b) - c * d >> state2`. Thus, there isn't really a way to make decisions based on the context without changing ownership of network components after the fact. One also has to keep in mind, that expressions can be used in multiple places, e.g. `tmp = a * b; tmp >> state1; tmp - c * d >> state2`.

It is also not necessarily clear to me how things should be grouped optimally if we had complete freedom. It might change depending on context. 

Addresses #126 and #127.


Let me know what you think of these things. Requested a review from @bjkomer too as he was intending to use the new SPA syntax himself.

**Interactions with other PRs:**
none

**How has this been tested?**
Looked at it in Nengo GUI.

**How long should this take to review?**
<!--- Please estimate if this PR is a quick, average, or lengthy PR. -->
<!--- Take into account both the size and complexity of the changes. -->
<!--- Also note if this is a timely PR that should be reviewed by a certain date. -->
<!--- Leave only the line that applies below: -->

- Average (neither quick nor lengthy)

**Where should a reviewer start?**
<!--- If the PR warrants it, indicate where a reviewer should start reviewing. -->
<!--- All lengthy PRs and complicated average PRs warrant this section. -->
<!--- If the PR is quick or straightforward, remove this section. -->

**Types of changes:**
<!--- What types of changes does your code introduce? -->
<!--- Leave all lines that apply below: -->

- New feature (non-breaking change which adds functionality)

**Checklist:**
<!--- Go over all the following points. Put an `x` in all the boxes that apply. -->
<!--- If a box is not applicable, please justify below the checklist. -->
<!--- If you're unsure about any of these, don't hesitate to ask. -->
<!--- We're here to help! -->

- [x] I have read the **CONTRIBUTING.rst** document.
- [n/a] I have updated the documentation accordingly.
- [n/a] I have included a changelog entry.
- [n/a] I have added tests to cover my changes.
- [x] I have run the test suite locally and all tests passed.

**Still to do:**
Further improvements? Discussion?

Seanny123 commented:
Nice! I have a few comments.

In the case where inputs to the network are known, would it be useful to add that information into the label? For example, in the code:

```
import nengo
import nengo_spa as spa

dim = 32
vocab = spa.Vocabulary(dimensions=dim)

words = ['ORANGE', 'APRICOT', 'CHERRY', 'STRAWBERRY', 'APPLE']

for word in words:
    vocab.populate(word)



with spa.Network('AssociativeMemory', seed=1) as model:
    a = spa.State(vocab, label='a')
    b = spa.State(vocab, label='b')
    
    model.assoc_mem = spa.ThresholdingAssocMem(threshold=0.3, input_vocab=vocab)
    a*b >> model.assoc_mem

```

Would it be helpful if the `Bind` network was named `Bind(a,b)`?

jgosmann commented:
> Would it be helpful if the Bind network was named Bind(a,b)?

Maybe? But it is also redundant information, might produce long names, and complicates the implementation.

Seanny123 commented:
> redundant information, might produce long names, and complicates the implementation.

Legit. I was thinking about all the interim populations I had created in other models and how I managed them. However, I think I should try building large networks in the new syntax and see how they display before commenting further.

jgosmann commented:
It might be worth to get this merged before the summer school?

Seanny123 commented:
To review this properly, I have to test it with the Nengo GUI. However, I'm a bit confused which version I should test against. Should I test against the last release or the Master? Is there going to be a release before the Summer School? I realise making a new release is being discussed in a [Nengo GUI issue](https://github.com/nengo/nengo-gui/issues/951).

jgosmann commented:
I think any “recent” release or development version of the GUI should be fine to test that. The networks should be displayed the same. If you want to have the SPA related plots, you will need nengo/nengo-gui#891.

Seanny123 commented:
This is definitely a huge improvement from before. However, as you mention, I think grouping all the connections inside of the Action Selection object, meaning it just looks kind of floating, is a bit confusing. I think we've talked about this before and you do seem to discuss it in the original PR text, but would you mind explaining in greater depth the different options? Maybe it's fine for the summer school and we can revisit how to visualize bg/thal connections later after the Nengo GUI refactor?

I wasn't really able to test the symbolic nodes. Could you give me some code showing them off so I could experiment with the feature?

jgosmann commented:
> I think grouping all the connections inside of the Action Selection object, meaning it just looks kind of floating, is a bit confusing. I think we've talked about this before and you do seem to discuss it in the original PR text, but would you mind explaining in greater depth the different options? Maybe it's fine for the summer school and we can revisit how to visualize bg/thal connections later after the Nengo GUI refactor?

I suppose one of the first question to ask (and I'm sure we discussed this before, but can't remember the details) whether the GUI should hide connections that are defined inside a network, but connect to things outside. If the answer is that the GUI should handle this differently this might become a non-issue for Nengo SPA, otherwise there are a few things to think about:

* Technically the connections are created within the ActionSelection network, so technically they are handled correctly. But what is technically correct might still be unintuitive or surprising to the user.
* It is not clear how to avoid the connections being part of the ActionSelection network:
  * We could delay the construction until the ActionSelection context is left, but this adds complexity as we need to keep track of the connections to add and might be surprising as everything else gets constructed immediately.
  * We could change the ownership of the connections after leaving the action selection context. But that might be surprising again.
  * We could potentially create the connections within a different context when they are created. Seems like the least surprising, but potentially still surprising.
  * ...?

In all of these cases I am not sure how the correct network to add things to could be determined (I'd need to look at the code in detail). And what network should they be added to? The parent of the ActionSelection network? (We don't have parent references in the networks ...) The network one above in the context stack? (Which tends to be the parent network, but might not always be the parent network.)

There is definitely some careful thinking required and to me it is not clear what the optimal solution would be even ignoring the implementation details.

Seanny123 commented:
I think part of the problem with connections is they communicate two different things to me:

1. Anatomical connections. Which networks are physically connected to each other.
2. Functional connections. Where the data flows in a network.

When it comes to the bg+thal, I want the connections to do both at the same time and I feel like that's not a good idea.

All the things you mentioned as "surprising" violate the first principle, but to me, the current state violates the second. Unsurprisingly, [we've discussed this before](https://github.com/nengo/nengo-gui/issues/922).

After talking to @jgosmann, we've decided:
- Forcing people to wrap the action selection results in a network to make a pretty layout is less baffling than hiding the connections completely.
- Later, we will be able to pursue more ideal cases such as:
  1. Forcing the action selection network to be open by default
  2. Having multiple views in the Nengo GUI to support large scale networks where you can differentiate the anatomical connections from the functional ones

@tcstewar do you have any opinion on this?

jgosmann commented:
@tcstewar Would you like to comment? It would be nice to get out another release with the changes in this PR before the summer school (and I am probably camping :tent: :evergreen_tree:  on the weekend, so it needs to happen tomorrow).

tcstewar commented:
> Forcing people to wrap the action selection results in a network to make a pretty layout is less baffling than hiding the connections completely.

That makes sense to me.  I like that reasoning.

> Later, we will be able to pursue more ideal cases

One possibility would be to adjust nengo_gui such that it looks inside closed Networks (but only one layer deep) and draws Connections it finds there.  That would solve this nicely, I think....

jgosmann commented:
I squashed the last commit and ammended it to remove the `hash` method that didn't have any purpose with the recent change. Also added another test, to meet the codecov criterion.

Still looks good?
